owasp-top10-2017-apps/a5/tictactoe/README.md

```diff
@@ -0,0 +1,271 @@
+
+  
+
+```sh
+
+cd secDevLabs/owasp-top10-2017-apps/a5/tictactoe
+
+```
+
+  
+
+```sh
+
+make install
+
+```
+
+  
+
+Then simply visit [localhost:10005][App] ! 😆
+
+  
+
+## Get to know the app 💵
+
+  
+
+To properly understand how this application works, you can follow this step:
+
+  
+
+- Try registering a user, sign in, play the game and see your statistics!
+
+  
+
+## Attack narrative
+
+  
+
+Now that you know the purpose of this app, what could go wrong? The following section describes how an attacker could identify and eventually find sensitive information about the app or its users. We encourage you to follow these steps and try to reproduce them on your own to better understand the attack vector! 😜
+
+  
+
+### 👀
+
+  
+
+In order to better understand how this app works, two users, `fry123` and `bender`, were created using web interface by visiting localhost:10005/create as shown below:
+
+  
+  
+
+<p  align="center">
+
+<img  src="images/attack0.png"/>
+
+</p>
+
+  
+
+#### Lack of user cookie validation allows for an attacker to get other users' game statistics
+
+  
+  
+
+After registering, the user `fry123` can sign in throught the route /login, start playing the game:
+
+  
+  
+
+<p  align="center">
+
+<img  src="images/attack1.png"/>
+
+</p>
+
+  
+
+And see his private statistics:
+
+  
+<p  align="center">
+
+<img  src="images/attack2.png"/>
+
+</p>
+
+  
+
+Using another user `bender` and the Burp Suite Tool, we can see the API request details when accessing _localhost:10005/statistics_. As seen below, to get the user statistics data, an ajax request is made passing the username as a URL parameter:
+
+  
+
+<p  align="center">
+
+<img  src="images/attack3.png"/>
+
+</p>
+
+
+And the request returns the user's statistics data:
+
+<p  align="center">
+
+<img  src="images/attack4.png"/>
+
+</p>
+
+  
+
+This may indicate that only this parameter is being used to verify the permission to get the data. So to check it, we will intercept the request and change the _user_ parameter by `fry123` using Burp Repeater. As we can see below, the data returned belongs to `fry123` user:
+
+  
+
+<p  align="center">
+
+<img  src="images/attack5.png"/>
+
+</p>
+
+  
+
+This allows an attacker to see every known user's private statistics.
+
+  
+
+#### Lack of user cookie validation allows for an attacker to manipulate user statistics
+
+  
+Analyzing the browser inspector we can see that every time a game ends, a request is made.
+
+  
+
+<p  align="center">
+
+<img  src="images/attack6.png"/>
+
+</p>
+
+  
+
+Using Burp Suite Tool, we can see the request details and notice that two parameters are sent, the username and the match result:
+
+  
+
+<p  align="center">
+
+<img  src="images/attack7.png"/>
+
+</p>
+
+  
+
+Using Burp Repeater we changed the _user_ parameter by `fry123` and, as shown below, it was  possible to add a new result to this user:
```

Comment:
What do you think of: 
`Sending this request to Burp Repeater and change the user parameter to another one, for example, fry123, an attacker could try to add a new result without the proper authorization, as can be seen in the following image:`

Reply:
I've changed it. Thank you for the suggestion, it is much better!